### PR TITLE
Multi-IdP for orgs: Data Model

### LINKED ISSUE #48907: Multi-IdP for orgs: Data Model
Implement the V3 data model for multi-IdP and multi-domain organization support.
  This covers two normalized data structures and the migration from the current production schema.

  ### New tables

  **`DOMAIN`** — global, realm-scoped domain registry (replaces the flat `ORG_DOMAIN` table):

  | Column | Type | Notes |
  |--------|------|-------|
  | ID | VARCHAR(36) | PK |
  | NAME | VARCHAR(255) | NOT NULL |
  | VERIFIED | BOOLEAN | Default false |
  | REALM_ID | VARCHAR(255) | NOT NULL; UK(REALM_ID, NAME) |
  | IDP_ID | VARCHAR(255) | Nullable FK → IDENTITY_PROVIDER.INTERNAL_ID |
  | AUTO_REDIRECT | BOOLEAN | Default false; replaces kc.org.broker.redirect.mode.email-matches IdP config |

  Domain-to-IdP routing is defined once globally on this table. This enforces by design
  that a shared domain always routes to the same IdP across all orgs (constraint C6).

  **`ORG_IDENTITY_PROVIDER`** — M:N join table between orgs and IdPs:

  | Column | Type | Notes |
  |--------|------|-------|
  | ORGANIZATION_ID | VARCHAR(255) | PK, FK → ORG.ID (CASCADE) |
  | IDENTITY_PROVIDER_ID | VARCHAR(255) | PK, FK → IDENTITY_PROVIDER.INTERNAL_ID (CASCADE) |
  | AUTO_MEMBERSHIP | BOOLEAN | Default true |
  | MEMBERSHIP_TYPE | VARCHAR(255) | Default 'UNMANAGED' |

  **`ORG_DOMAIN`** — reshaped from flat table to M:N join table (DOMAIN_ID, ORG_ID).
  Same name, different shape. Allows the same domain to be claimed by multiple orgs.

  ### Removed structures

  - `IDENTITY_PROVIDER.ORGANIZATION_ID` column (replaced by `ORG_IDENTITY_PROVIDER` join table)
  - `kc.org.domain` and `kc.org.excluded.domains` IdP config entries 
    - replaced by per-domain explicit configuration
      - Each DOMAIN row has its own IDP_ID (nullable) and AUTO_REDIRECT boolean
      - To achieve the old "all except X" behavior, set IDP_ID on the domains you want routed and leave it NULL on the ones you don't
      - There's no need for an exclusion list because each domain independently declares its routing
  - `kc.org.broker.redirect.mode.email-matches` IdP config (replaced by `DOMAIN.AUTO_REDIRECT`)
  - `OrganizationModel.ORGANIZATION_DOMAIN_ATTRIBUTE` and `ORGANIZATION_EXCLUDED_DOMAIN_ATTRIBUTE` constants
  - `OrganizationModel.IdentityProviderRedirectMode` enum

  ### SPI changes
  
  - `OrganizationDomainModel` extended with `identityProviderAlias` and `autoRedirect` fields
  - `OrganizationDomainRepresentation` updated to match

### Migration

- Schema updates
- migration from IdpConfig entries to new structure
- migration for exported JSONs while importing (`migrateImport`)

### PR DESCRIPTION
**Pull request by vramik**
Closes #48907

### CODE REVIEW

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_PopulateOrgDomainJoinMigration.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        String orgDomainOldTable = database.correctObjectName("ORG_DOMAIN_OLD", Table.class);
+        String domainTable = database.correctObjectName("DOMAIN", Table.class);
+        String orgDomainTable = database.correctObjectName("ORG_DOMAIN", Table.class);
+
+        try (PreparedStatement ps = connection.prepareStatement(
+                "SELECT dom.ID, od.ORG_ID FROM " + getTableName(orgDomainOldTable) +
+                        " od JOIN " + getTableName(domainTable) + " dom ON dom.NAME = od.NAME");
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_PopulateOrgDomainJoinMigration.java]: updated

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+import liquibase.statement.core.InsertStatement;
+import liquibase.statement.core.UpdateStatement;
+import liquibase.structure.core.Table;
+
+public class JpaUpdate26_8_0_DomainIdpRoutingMigration extends CustomKeycloakTask {
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        populateOrgIdentityProvider();
+        populateDomainIdpRouting();
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java:37]: updated

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+                boolean autoRedirect = isEmailMatchEnabled(idpInternalId);
+
+                if (!"ANY".equals(domainValue)) {
+                    // Case A: specific domain name
+                    statements.add(new UpdateStatement(null, null, domainTable)
+                            .addNewColumnValue("IDP_ID", idpInternalId)
+                            .addNewColumnValue("AUTO_REDIRECT", autoRedirect)
+                            .setWhereClause("NAME=?")
+                            .addWhereParameter(domainValue));
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java]: updated

File: model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java [hunk 4]
```diff
@@ -31,4 +45,64 @@ public void migrateRealm(KeycloakSession session, RealmModel realm) {
...
+
+                boolean autoRedirect = Boolean.parseBoolean(config.get(MigrationUtils.ORGANIZATION_REDIRECT_MODE_ATTRIBUTE));
+
+                Set<OrganizationDomainModel> result = new HashSet<>();
+                for (OrganizationDomainModel domain : currentDomains) {
+                    if ("ANY".equals(orgDomain)) {
+                        if (!excludedDomains.contains(domain.getName())) {
+                            result.add(new OrganizationDomainModel(domain.getName(), domain.isVerified(), idp.getAlias(), autoRedirect));
+                        } else {
+                            result.add(domain);
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java:90]: `ANY` and exclusion concept is being replaced by direct relationship between domain and IDP,

File: model/jpa/src/main/java/org/keycloak/organization/jpa/OrganizationAdapter.java [hunk 5]
```diff
@@ -182,36 +186,45 @@ public void setDomains(Set<OrganizationDomainModel> domains) {
...
-                getIdentityProviders()
-                        .filter(idp -> Objects.equals(domainEntity.getName(), idp.getConfig().get(ORGANIZATION_DOMAIN_ATTRIBUTE)))
-                        .forEach(idp -> {
-                            idp.getConfig().remove(ORGANIZATION_DOMAIN_ATTRIBUTE);
-                            session.identityProviders().update(idp);
-                        });
+                domainEntity.setIdentityProvider(null);
+                // TODO: when separate domain management allows M:N sharing, only remove the join row
+                //  and rely on orphan cleanup instead of deleting the DOMAIN row outright
+                em.remove(domainEntity);
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/OrganizationAdapter.java:207]: The TODO comment in the code explicitly notes this. Shared domain lifecycle needs proper handling when separate domain management lands, part of different issue

File: model/jpa/src/main/java/org/keycloak/organization/jpa/OrganizationAdapter.java [hunk 6]
```diff
@@ -182,36 +186,45 @@ public void setDomains(Set<OrganizationDomainModel> domains) {
...
-            domainEntity.setId(KeycloakModelUtils.generateId());
-            domainEntity.setName(model.getName());
-            domainEntity.setVerified(model.isVerified());
-            domainEntity.setOrganization(this.entity);
+            OrganizationDomainEntity domainEntity;
+            try {
+                domainEntity = em.createNamedQuery("getDomainByRealmAndName", OrganizationDomainEntity.class)
+                        .setParameter("realmId", realm.getId())
+                        .setParameter("name", model.getName())
+                        .getSingleResult();
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/OrganizationAdapter.java:217]: Planned to be fixed as part of #48908

File: model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java [hunk 7]
```diff
@@ -702,16 +709,25 @@ public boolean addIdentityProvider(OrganizationModel organization, IdentityProvi
...
+                .anyMatch(oip -> oip.getIdentityProviderId().equals(identityProvider.getInternalId()));
+        if (alreadyLinked) {
             return false;
-        } else if (orgId != null) {
+        }
+
+        // TODO: remove this 1:1 check when M:N is fully supported
+        String existingOrgId = identityProvider.getOrganizationId();
+        if (existingOrgId != null) {
             throw new ModelValidationException("Identity provider already associated with a different organization");
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/JpaOrganizationProvider.java:721]: Planned to be fixed as part of #48908

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/OrganizationEntity.java [hunk 8]
```diff
@@ -40,7 +43,7 @@
 @Entity
 @NamedQueries({
         @NamedQuery(name="getByOrgName", query="select distinct o from OrganizationEntity o where o.realmId = :realmId AND o.name = :name"),
-        @NamedQuery(name="getByDomainName", query="select distinct o from OrganizationEntity o inner join OrganizationDomainEntity d ON o.id = d.organization.id" +
+        @NamedQuery(name="getByDomainName", query="select distinct o from OrganizationEntity o inner join o.domains d" +
                 " where o.realmId = :realmId and d.name in (:names)"),
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/OrganizationEntity.java:47]: Planned to be fixed as part of #48908

File: server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java [hunk 9]
```diff
@@ -1893,13 +1892,9 @@ private static void updateOrganizationBroker(IdentityProviderRepresentation repr
...
-            String domain = representation.getConfig().get(OrganizationModel.ORGANIZATION_DOMAIN_ATTRIBUTE);
-
-            if (StringUtil.isBlank(domain)) {
-                representation.getConfig().remove(OrganizationModel.ORGANIZATION_DOMAIN_ATTRIBUTE);
-            } else if (!ANY_DOMAIN.equals(domain) && org.getDomains().map(OrganizationDomainModel::getName).noneMatch(domain::equals)) {
-                throw new IllegalArgumentException("Domain does not match any domain from the organization");
-            }
+            // strip old domain config entries that are no longer used
+            representation.getConfig().remove(MigrationUtils.ORGANIZATION_DOMAIN_ATTRIBUTE);
+            representation.getConfig().remove(MigrationUtils.ORGANIZATION_EXCLUDED_DOMAIN_ATTRIBUTE);
```
[the remark this answers is not in the document]
Comment (vramik) [reply server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java:1901]: MOSTLY FALSE POSITIVE. MigrateTo26_8_0.doMigrateDomainIdpRouting() (line 103) already strips all three legacy config keys including ORGANIZATION_REDIRECT_MODE_ATTRIBUTE. Minor gap: updateOrganizationBroker could strip it defensively for import-only paths, but not a functional bug, adding it anyway

[the remark this answers is not in the document]
Comment (vramik) [reply server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java:1901]: updated

File: model/jpa/src/main/java/org/keycloak/organization/jpa/OrganizationAdapter.java [hunk 10]
```diff
@@ -297,6 +312,17 @@ private OrganizationDomainModel validateDomain(OrganizationDomainModel domainMod
...
 
+    private IdentityProviderEntity resolveIdentityProvider(EntityManager em, String alias) {
+        if (alias == null) {
+            return null;
+        }
+        IdentityProviderModel idpModel = session.identityProviders().getByAlias(alias);
+        if (idpModel == null) {
+            throw new ModelValidationException("Identity provider with alias '" + alias + "' does not exist in realm " + realm.getName());
+        }
+        return em.getReference(IdentityProviderEntity.class, idpModel.getInternalId());
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/organization/jpa/OrganizationAdapter.java]: updated

File: model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java [hunk 11]
```diff
@@ -31,4 +45,75 @@ public void migrateRealm(KeycloakSession session, RealmModel realm) {
...
+                        excludedDomains.add(d.trim());
+                    }
+                }
+
+                boolean autoRedirect = Boolean.parseBoolean(config.get(MigrationUtils.ORGANIZATION_REDIRECT_MODE_ATTRIBUTE));
+
+                Set<OrganizationDomainModel> result = new HashSet<>();
+                for (OrganizationDomainModel domain : currentDomains) {
+                    if ("ANY".equals(orgDomain)) {
+                        if (!excludedDomains.contains(domain.getName())) {
```
Comment (martin-kanis) [inline model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java:87]: No wildcard pattern matching - this will not correctly handle wildcard exclusions like `*.example.com`

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+                            "SELECT dom.ID, dom.NAME FROM " + getTableName(domainTable) + " dom" +
+                                    " JOIN " + getTableName(orgDomainTable) + " od ON od.DOMAIN_ID = dom.ID" +
+                                    " WHERE od.ORG_ID = ? AND dom.IDP_ID IS NULL")) {
+                        domPs.setString(1, orgId);
+                        try (ResultSet domRs = domPs.executeQuery()) {
+                            while (domRs.next()) {
+                                String domainId = domRs.getString(1);
+                                String domainName = domRs.getString(2);
+                                existingDomainNames.add(domainName);
+                                if (!excludedDomains.contains(domainName)) {
```
Comment (martin-kanis) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java:184]: No wildcard pattern matching - this will not correctly handle wildcard exclusions like `*.example.com`

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+import java.util.Set;
+import java.util.UUID;
+
+import liquibase.exception.CustomChangeException;
+import liquibase.statement.core.DeleteStatement;
+import liquibase.statement.core.InsertStatement;
+import liquibase.statement.core.UpdateStatement;
+import liquibase.structure.core.Table;
+
+public class JpaUpdate26_8_0_DomainIdpRoutingMigration extends CustomKeycloakTask {
```
Comment (martin-kanis) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java:32]: This migration is quite a complex one. Should we introduce some tests or are we confident enough that it will work fine across many possible edge-cases and large datasets?

File: model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java [hunk 14]
```diff
@@ -1,12 +1,20 @@
...
 import org.keycloak.migration.MigrationProvider;
 import org.keycloak.migration.ModelVersion;
 import org.keycloak.models.ClientScopeModel;
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.OrganizationDomainModel;
 import org.keycloak.models.RealmModel;
+import org.keycloak.organization.OrganizationProvider;
+import org.keycloak.representations.idm.RealmRepresentation;
 
 public class MigrateTo26_8_0 extends RealmMigration {
```
Comment (martin-kanis) [inline model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java:19]: Should we mention this in the migration guide or it will be done as a follow-up?

**Review by martin-kanis — COMMENTED** [review]
@vramik Thanks for the PR. 
I did the first round of review that focuses mostly on the migration part. I have some comments/concerns. Please see below. 

In the meantime, I'll continue with the review and I might have more comments later.

File: model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java [hunk 15]
```diff
@@ -31,4 +45,75 @@ public void migrateRealm(KeycloakSession session, RealmModel realm) {
...
+                    return;
+                }
+
+                Set<OrganizationDomainModel> currentDomains = org.getDomains().collect(Collectors.toSet());
+                String excludedDomainsValue = config.get(MigrationUtils.ORGANIZATION_EXCLUDED_DOMAIN_ATTRIBUTE);
+                Set<String> excludedDomains = new HashSet<>();
+                if (excludedDomainsValue != null) {
+                    for (String d : excludedDomainsValue.split(",")) {
+                        excludedDomains.add(d.trim());
+                    }
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java:79]: Domain names are stored as-is — there's no normalization to lowercase on creation. The `toLowerCase()` calls in `Organizations.java`  are only used in the domain-matching utility (`matchesDomain()`), not when storing.

File: model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java (hunk 14 above)
Comment (vramik) [reply model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java:19]: We will document the new capability in release notes, perhaps there will be a note in migration guide needed, especially about moving configs from IDP config table elsewhere, but I'd prefer to do it in one go and not within this one.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java (hunk 13 above)
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java:32]: Added tests to `MigrationTest`.

File: model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java (hunk 11 above)
Comment (vramik) [reply model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_8_0.java:87]: As discussed via DM, the exclusions (even the wildcard ones) should be represented by an entry in `DOMAIN` table with `NULL` value in `IDP_ID` column. `MigrationTest` should verify it.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java (hunk 12 above)
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_8_0_DomainIdpRoutingMigration.java:184]: As discussed via DM, the exclusions (even the wildcard ones) should be represented by an entry in `DOMAIN` table with `NULL` value in `IDP_ID` column. `MigrationTest` should verify it.

File: core/src/main/java/org/keycloak/representations/idm/OrganizationDomainRepresentation.java [hunk 16]
```diff
@@ -32,6 +32,8 @@ public class OrganizationDomainRepresentation {
 
     private String name;
     private boolean verified;
+    private String identityProviderAlias;
+    private boolean autoRedirect;
```
[the remark this answers is not in the document]
Comment (vramik) [reply core/src/main/java/org/keycloak/representations/idm/OrganizationDomainRepresentation.java:36]: Removed the configs from UI for now

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaIdentityProviderStorageProvider.java [hunk 17]
```diff
@@ -579,7 +601,7 @@ private IdentityProviderModel toModel(IdentityProviderEntity entity) {
         identityProviderModel.setAuthenticateByDefault(entity.isAuthenticateByDefault());
         identityProviderModel.setFirstBrokerLoginFlowId(entity.getFirstBrokerLoginFlowId());
         identityProviderModel.setPostBrokerLoginFlowId(entity.getPostBrokerLoginFlowId());
-        identityProviderModel.setOrganizationId(entity.getOrganizationId());
+        identityProviderModel.setOrganizationId(getOrganizationIdForIdp(entity.getInternalId()));
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaIdentityProviderStorageProvider.java:604]: It's intermediate state, it should be handled in follow-up issues

**Comment by vramik** [comment]
Closing in fovor of #52401
